docs/tutorials/handling-files.rst

```diff
@@ -0,0 +1,591 @@
+images using the IIIF de facto standard.
+
+.. _handling-files-overview:
+
+Overview
+--------
+The first step when setting up files integration is to decide where to store them.
+In Invenio, you can achieve this by defining a ``Location``: a location
+is a representation of your storage system. It can be, for example, a local
+folder in your machine or a remote storage. It has a ``name`` and a ``URI`` to
+define the base path to the files.
+
+To access and manage files in a location, Invenio uses ``Storage`` implementation:
+it defines how to access files in your location. In other words, it knows how files
+are stored, e.g. in a hierarchy of folders, and how to read and write them.
+
+You can define multiple locations and storage. This is useful, for example,
+when you are dealing with online and offline file systems.
+
+A single specific file is represented in Invenio by a ``FileInstance`` object:
+it defines its relative path to the Location where it is stored and also
+other useful properties such as size or checksum.
+
+In a few words, you can think to Location, Storage and FileInstance as
+the physical representation of your files system(s).
+
+Object storage
+++++++++++++++
+Invenio provides an abstraction of your physical file system with an object
+storage representation similar to Amazon S3. This allows great flexibility
+and easy to use APIs.
+
+With an analogy to the traditional file systems where files are
+contained in folders, in Invenio object storage ``ObjectVersions``
+are contained in ``Buckets``.
+
+Buckets are a set of objects, they are uniquely identified by an ID
+and can define constraints on file sizes or quotas. A bucket also
+defines the default Location to use when adding files, if none provided.
+
+An ObjectVersion is a specific version (the ``version_id``) of a file
+at a given moment in time. It has a reference to a FileInstance
+and metadata such as the file name (the ``key``). The latest version
+of a file is marked as the ``head``.
+
+ObjectVersions allow to perform operations to your files without
+accessing the file system. Let's see a common example: a user wants
+to delete a previously uploaded file.
+With Invenio, this means creating a new ObjectVersion for this
+file, the new head, without reference to the FileInstance that
+it had before. The file on disk is not accessed or removed.
+This is also called delete marker or soft deletion.
+
+You can read more in the
+`Invenio-Files-REST <http://invenio-files-rest.readthedocs.io/en/latest/>`__
+module documentation.
+
+.. _handling-files-integration-with-records:
+
+Integration with records
+------------------------
+Invenio integrates files and records by creating a reference
+between a Record and a Bucket. Usually a record has a reference to one
+bucket and a bucket to one record.
+
+.. image:: ../_static/invenio-files-integration.png
+
+Invenio allows you to set up different scenarios and have, for example,
```

Comment:
Maybe we need to specify that this behavior does not come out of the box and that the user needs to implement their own code to and override parts of the existing class to achieve this.

Reply:
Right, I have added `To achieve any of these, you will have to define
your own integration between records and files.`